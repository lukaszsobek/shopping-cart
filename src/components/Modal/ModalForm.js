import React, { Component } from "react";
import { connect } from "react-redux";

import Button from "../shared/Button";

import { addItem, closeModal } from "../../actions";

const ConstructDropdown = (props) => {

    const listOptions = props.productList.map((item, key) => {
        return (
            <option
                value={ key }
                key={ key }
            >{ item.name }</option>
        );
    })

    return (
        <select
            name={ props.name }
            className={ props.className }
        >
            {listOptions}
        </select>
    );
}

class ModalForm extends Component {

    handleSubmit(e) {
        e.preventDefault();

        const { addItem, closeModal, availableProducts } = this.props;
        const { productComment, productDropdown }  = e.target.elements;

        const productIndex = productDropdown[productDropdown.selectedIndex].value;
        const newItem = {
            ...availableProducts[productIndex],
            comments: productComment.value
        }

        addItem(newItem);
        closeModal();

    }

    render() {
        const { availableProducts } = this.props;

        return (
            <form
                id="modal-panel-form"
                name="modalPanelForm"
                onSubmit={ e => this.handleSubmit(e) }
            >
                <ConstructDropdown
                    productList={ availableProducts }
                    name="productDropdown"
                    className="product-dropdown"
                />

                <div className="input-wrapper">
                    <label>Comment:</label><br/>
                    <input
                        name="productComment"
                        className="product-comment"
                    />
                </div>

                <Button
                    className="add-item"
                    label="Add to Cart"
                />  
            </form>
        )
    }
}

const mapStateToProps = state => ({
    availableProducts: state.availableProducts
});

const mapDispatchToProps = dispatch => ({
    addItem: item => dispatch(addItem(item)),
    closeModal: () => dispatch(closeModal())
})

export default connect(mapStateToProps,mapDispatchToProps)(ModalForm);