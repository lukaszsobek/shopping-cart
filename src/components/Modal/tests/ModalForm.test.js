import { ConstructDropdown, ModalForm } from "../ModalForm";

describe("ModalForm", () => {
    it("constructs dropdown", () => {
        const props = {
            productList: [{ name: "test1" }, {name: "test2"}],
            name: "testName",
            className: "testClass"
        }
        expect(ConstructDropdown(props)).toMatchSnapshot();
    });

    it("renders", () => {
        const props = {
            availableProducts: [{ name: "test1" }, {name: "test2"}]
        }
        const testModalForm = new ModalForm(props).render();
        expect(testModalForm).toMatchSnapshot();
    })
});

