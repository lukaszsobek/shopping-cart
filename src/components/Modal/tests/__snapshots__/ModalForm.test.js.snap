// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalForm constructs dropdown 1`] = `
<select
  className="testClass"
  name="testName"
>
  <option
    value={0}
  >
    test1
  </option>
  <option
    value={1}
  >
    test2
  </option>
</select>
`;

exports[`ModalForm renders 1`] = `
<form
  id="modal-panel-form"
  name="modalPanelForm"
  onSubmit={[Function]}
>
  <h3>
    Update item
  </h3>
  <div
    className="product-selection-wrapper"
  >
    <label
      htmlFor="productDropdown"
    >
      Product:
    </label>
    <br />
    <Unknown
      className="product-dropdown"
      id="productDropdown"
      name="productDropdown"
      productList={
        Array [
          Object {
            "name": "test1",
          },
          Object {
            "name": "test2",
          },
        ]
      }
      selectedItem={undefined}
    />
  </div>
  <div
    className="product-comment-wrapper"
  >
    <label
      htmlFor="productComment"
    >
      Comment:
    </label>
    <br />
    <input
      className="product-comment"
      id="productComment"
      name="productComment"
    />
  </div>
  <Button
    className="add-item-button"
    label="Update item"
  />
</form>
`;
